File: src/ai_engine.rs
use std::collections::HashMap;

use rand::Rng;

use cocos4_rust::base::value::{Value, ValueMap};

use crate::atom::{AtomId, AtomRegistry};
use crate::gameplay::GameplayType;

#[derive(Debug, Clone)]
pub struct GenerationConfig {
    pub min_atoms: usize,
    pub max_atoms: usize,
    pub difficulty_range: (f32, f32),
    pub allow_composite: bool,
    pub seed: Option<u64>,
    pub player_level: u32,
    pub preferred_types: Vec<GameplayType>,
    pub excluded_types: Vec<GameplayType>,
    pub reward_multiplier: f32,
}

impl Default for GenerationConfig {
    fn default() -> Self {
        Self {
            min_atoms: 2,
            max_atoms: 4,
            difficulty_range: (0.5, 1.0),
            allow_composite: true,
            seed: None,
            player_level: 1,
            preferred_types: Vec::new(),
            excluded_types: Vec::new(),
            reward_multiplier: 1.0,
        }
    }
}

impl GenerationConfig {
    pub fn for_player_level(level: u32) -> Self {
        let max_atoms = (2 + level / 5).min(6) as usize;
        let difficulty = 0.3 + (level as f32 * 0.1).min(0.7);
        Self {
            min_atoms: 2,
            max_atoms,
            difficulty_range: (difficulty, difficulty + 0.3),
            player_level: level,
            ..Default::default()
        }
    }
}

#[derive(Debug, Clone)]
pub struct DimensionBlueprint {
    pub id: String,
    pub name: String,
    pub description: String,
    pub atom_ids: Vec<AtomId>,
    pub atom_weights: HashMap<AtomId, f32>,
    pub difficulty: f32,
    pub rules: Vec<GeneratedRule>,
    pub rewards: Vec<GeneratedReward>,
    pub theme: DimensionTheme,
    pub time_limit_secs: Option<u32>,
    pub objectives: Vec<Objective>,
}

#[derive(Debug, Clone)]
pub struct GeneratedRule {
    pub rule_id: String,
    pub name: String,
    pub description: String,
    pub rule_type: RuleType,
    pub params: ValueMap,
}

#[derive(Debug, Clone)]
pub enum RuleType {
    Modifier,
    Constraint,
    Trigger,
    Transformation,
}

#[derive(Debug, Clone)]
pub struct GeneratedReward {
    pub item_id: String,
    pub base_quantity: u32,
    pub scaling_factor: f32,
}

#[derive(Debug, Clone)]
pub struct DimensionTheme {
    pub name: String,
    pub visual_style: String,
    pub music_mood: String,
    pub color_palette: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct Objective {
    pub id: String,
    pub description: String,
    pub objective_type: ObjectiveType,
    pub target_value: u64,
    pub is_optional: bool,
}

#[derive(Debug, Clone)]
pub enum ObjectiveType {
    Score,
    Time,
    Collect,
    Defeat,
    Survive,
    Custom(String),
}

pub struct DimensionGenerator {
    rng: rand::rngs::StdRng,
    name_parts: NameParts,
}

struct NameParts {
    adjectives: Vec<&'static str>,
    nouns: Vec<&'static str>,
    themes: Vec<&'static str>,
}

impl DimensionGenerator {
    pub fn new(seed: u64) -> Self {
        Self {
            rng: rand::SeedableRng::seed_from_u64(seed),
            name_parts: NameParts {
                adjectives: vec![
                    "混沌", "永恒", "幻影", "量子", "虚空", "烈焰", "冰霜",
                    "雷霆", "暗影", "光辉", "深渊", "星辰", "时空", "命运",
                ],
                nouns: vec![
                    "迷宫", "战场", "神殿", "深渊", "花园", "塔楼", "竞技场",
                    "秘境", "次元", "试炼", "回廊", "领域", "裂隙", "梦境",
                ],
                themes: vec![
                    "赛博朋克", "奇幻森林", "海底世界", "太空站", "古墓",
                    "浮空城", "熔岩地带", "冰原", "沙漠绿洲", "暗黑地牢",
                ],
            },
        }
    }

    pub fn generate(&mut self, config: &GenerationConfig, registry: &AtomRegistry) -> DimensionBlueprint {
        let available_atoms = self.filter_atoms(config, registry);
        let num_atoms = self.rng.gen_range(config.min_atoms..=config.max_atoms.min(available_atoms.len()));

        let selected_atoms = self.select_atoms(&available_atoms, num_atoms);
        let difficulty = self.rng.gen_range(config.difficulty_range.0..=config.difficulty_range.1);
        let atom_weights = self.generate_weights(&selected_atoms);
        let rules = self.generate_rules(&selected_atoms, difficulty);
        let rewards = self.generate_rewards(&selected_atoms, difficulty, config.reward_multiplier);
        let theme = self.generate_theme();
        let name = self.generate_name();
        let objectives = self.generate_objectives(&selected_atoms, difficulty);

        DimensionBlueprint {
            id: format!("dim_{}", self.rng.gen::<u32>()),
            name,
            description: format!("在{}中挑战{}种玩法的组合", theme.name, selected_atoms.len()),
            atom_ids: selected_atoms,
            atom_weights,
            difficulty,
            rules,
            rewards,
            theme,
            time_limit_secs: if difficulty > 0.7 { Some(180) } else { None },
            objectives,
        }
    }

    fn filter_atoms(&self, config: &GenerationConfig, registry: &AtomRegistry) -> Vec<AtomId> {
        registry
            .list_all()
            .iter()
            .filter(|m| {
                let gt = GameplayType::from_name(&m.gameplay_type);
                !config.excluded_types.contains(&gt)
            })
            .map(|m| m.id.clone())
            .collect()
    }

    fn select_atoms(&mut self, available: &[AtomId], count: usize) -> Vec<AtomId> {
        let mut indices: Vec<usize> = (0..available.len()).collect();
        let mut selected = Vec::new();

        for _ in 0..count.min(available.len()) {
            if indices.is_empty() {
                break;
            }
            let idx = self.rng.gen_range(0..indices.len());
            let atom_idx = indices.remove(idx);
            selected.push(available[atom_idx].clone());
        }

        selected
    }

    fn generate_weights(&mut self, atoms: &[AtomId]) -> HashMap<AtomId, f32> {
        let mut weights = HashMap::new();
        let total = atoms.len() as f32;
        for atom in atoms {
            let w = 0.5 + self.rng.gen::<f32>() * 0.5;
            weights.insert(atom.clone(), w / total);
        }
        weights
    }

    fn generate_rules(&mut self, _atoms: &[AtomId], difficulty: f32) -> Vec<GeneratedRule> {
        let mut rules = Vec::new();
        let rule_templates = [
            ("speed_boost", "加速", "行动速度提升", RuleType::Modifier),
            ("double_score", "双倍得分", "得分翻倍", RuleType::Modifier),
            ("resource_drain", "资源消耗", "资源持续消耗", RuleType::Constraint),
            ("chain_bonus", "连锁奖励", "连续操作获得额外奖励", RuleType::Trigger),
            ("time_pressure", "时间压力", "倒计时加速", RuleType::Constraint),
        ];

        let num_rules = (1 + (difficulty * 3.0) as usize).min(rule_templates.len());
        let mut template_indices: Vec<usize> = (0..rule_templates.len()).collect();

        for _ in 0..num_rules {
            if template_indices.is_empty() {
                break;
            }
            let idx = self.rng.gen_range(0..template_indices.len());
            let ti = template_indices.remove(idx);
            let template = &rule_templates[ti];
            let mut params = ValueMap::new();
            params.insert("intensity".to_string(), Value::Float(difficulty as f64));

            rules.push(GeneratedRule {
                rule_id: template.0.to_string(),
                name: template.1.to_string(),
                description: template.2.to_string(),
                rule_type: template.3.clone(),
                params,
            });
        }

        rules
    }

    fn generate_rewards(&mut self, _atoms: &[AtomId], difficulty: f32, multiplier: f32) -> Vec<GeneratedReward> {
        let mut rewards = Vec::new();
        let base_gold = (50.0 * difficulty * multiplier) as u32;
        let base_gem = (5.0 * difficulty * multiplier) as u32;

        rewards.push(GeneratedReward {
            item_id: "gold".to_string(),
            base_quantity: base_gold,
            scaling_factor: 1.0 + difficulty,
        });
        rewards.push(GeneratedReward {
            item_id: "gem".to_string(),
            base_quantity: base_gem.max(1),
            scaling_factor: 0.5 + difficulty,
        });

        if difficulty > 0.6 {
            rewards.push(GeneratedReward {
                item_id: "rare_chest".to_string(),
                base_quantity: 1,
                scaling_factor: difficulty,
            });
        }

        rewards
    }

    fn generate_theme(&mut self) -> DimensionTheme {
        let adj_idx = self.rng.gen_range(0..self.name_parts.adjectives.len());
        let theme_idx = self.rng.gen_range(0..self.name_parts.themes.len());

        DimensionTheme {
            name: format!("{}·{}", self.name_parts.adjectives[adj_idx], self.name_parts.themes[theme_idx]),
            visual_style: self.name_parts.themes[theme_idx].to_string(),
            music_mood: if self.rng.gen::<f32>() > 0.5 { "epic" } else { "mysterious" }.to_string(),
            color_palette: vec!["#FF6B6B".to_string(), "#4ECDC4".to_string(), "#45B7D1".to_string()],
        }
    }

    fn generate_name(&mut self) -> String {
        let adj_idx = self.rng.gen_range(0..self.name_parts.adjectives.len());
        let noun_idx = self.rng.gen_range(0..self.name_parts.nouns.len());
        format!("{}{}", self.name_parts.adjectives[adj_idx], self.name_parts.nouns[noun_idx])
    }

    fn generate_objectives(&mut self, atoms: &[AtomId], difficulty: f32) -> Vec<Objective> {
        let mut objectives = Vec::new();

        objectives.push(Objective {
            id: "main_score".to_string(),
            description: format!("达到{}分", (1000.0 * difficulty) as u64),
            objective_type: ObjectiveType::Score,
            target_value: (1000.0 * difficulty) as u64,
            is_optional: false,
        });

        if atoms.len() > 1 {
            objectives.push(Objective {
                id: "combo_master".to_string(),
                description: "完成3次组合连击".to_string(),
                objective_type: ObjectiveType::Custom("combo".to_string()),
                target_value: 3,
                is_optional: true,
            });
        }

        objectives
    }
}

pub struct BalanceTuner {
    target_win_rate: f32,
    history: Vec<BalanceDataPoint>,
}

#[derive(Debug, Clone)]
struct BalanceDataPoint {
    dimension_id: String,
    difficulty: f32,
    player_level: u32,
    score: u64,
    duration_secs: f32,
    completed: bool,
}

impl BalanceTuner {
    pub fn new() -> Self {
        Self {
            target_win_rate: 0.6,
            history: Vec::new(),
        }
    }

    pub fn record_result(&mut self, dimension_id: &str, difficulty: f32, player_level: u32, score: u64, duration_secs: f32, completed: bool) {
        self.history.push(BalanceDataPoint {
            dimension_id: dimension_id.to_string(),
            difficulty,
            player_level,
            score,
            duration_secs,
            completed,
        });
    }

    pub fn suggest_difficulty(&self, player_level: u32) -> f32 {
        let recent: Vec<_> = self.history
            .iter()
            .rev()
            .take(20)
            .filter(|d| (d.player_level as i32 - player_level as i32).abs() <= 2)
            .collect();

        if recent.is_empty() {
            return 0.3 + player_level as f32 * 0.05;
        }

        let win_rate = recent.iter().filter(|d| d.completed).count() as f32 / recent.len() as f32;

        let mut adjustment = 0.0f32;
        if win_rate > self.target_win_rate + 0.1 {
            adjustment += 0.1;
        } else if win_rate < self.target_win_rate - 0.1 {
            adjustment -= 0.1;
        }

        let base = 0.3 + player_level as f32 * 0.05;
        (base + adjustment).clamp(0.1, 1.0)
    }
}

impl Default for BalanceTuner {
    fn default() -> Self {
        Self::new()
    }
}

pub struct AiEngine {
    pub generator: DimensionGenerator,
    pub tuner: BalanceTuner,
}

impl AiEngine {
    pub fn new(seed: u64) -> Self {
        Self {
            generator: DimensionGenerator::new(seed),
            tuner: BalanceTuner::new(),
        }
    }

    pub fn generate_dimension(&mut self, config: &GenerationConfig, registry: &AtomRegistry) -> DimensionBlueprint {
        let mut config = config.clone();
        let suggested = self.tuner.suggest_difficulty(config.player_level);
        config.difficulty_range = (
            (suggested - 0.1).max(0.1),
            (suggested + 0.1).min(1.0),
        );

        self.generator.generate(&config, registry)
    }

    pub fn record_session(&mut self, dimension_id: &str, difficulty: f32, player_level: u32, score: u64, duration_secs: f32, completed: bool) {
        self.tuner.record_result(dimension_id, difficulty, player_level, score, duration_secs, completed);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::atom::AtomMetadata;

    fn make_test_registry() -> AtomRegistry {
        let mut registry = AtomRegistry::new();
        let atoms = vec![
            ("match3", "Match3", "match3"),
            ("tower_defense", "TowerDefense", "tower_defense"),
            ("card", "Card", "card"),
        ];

        for (id, name, gt) in atoms {
            let metadata = AtomMetadata {
                id: id.to_string(),
                name: name.to_string(),
                version: 1,
                gameplay_type: gt.to_string(),
                description: format!("{} atom", name),
                tags: vec![gt.to_string()],
            };
            registry.register(id.to_string(), metadata, || panic!("factory not used in test"));
        }
        registry
    }

    #[test]
    fn test_dimension_generator() {
        let mut gen = DimensionGenerator::new(42);
        let config = GenerationConfig::default();
        let registry = make_test_registry();

        let blueprint = gen.generate(&config, &registry);
        assert!(!blueprint.atom_ids.is_empty());
        assert!(!blueprint.name.is_empty());
    }

    #[test]
    fn test_ai_engine_generate() {
        let mut engine = AiEngine::new(123);
        let config = GenerationConfig::default();
        let registry = make_test_registry();

        let blueprint = engine.generate_dimension(&config, &registry);
        assert!(!blueprint.atom_ids.is_empty());
        assert!(!blueprint.objectives.is_empty());
    }
}

File: src/ai/AIEngine.ts
export interface GenerationConfig {
    minAtoms: number;
    maxAtoms: number;
    difficultyRange: [number, number];
    playerLevel: number;
    preferredTypes: string[];
    excludedTypes: string[];
    rewardMultiplier: number;
}

export interface DimensionBlueprint {
    id: string;
    name: string;
    description: string;
    atomIds: string[];
    atomWeights: Record<string, number>;
    difficulty: number;
    rules: GeneratedRule[];
    rewards: GeneratedReward[];
    theme: DimensionTheme;
    timeLimitSecs: number | null;
    objectives: Objective[];
}

export interface GeneratedRule {
    ruleId: string;
    name: string;
    description: string;
    ruleType: 'modifier' | 'constraint' | 'trigger' | 'transformation';
    params: Record<string, any>;
}

export interface GeneratedReward {
    itemId: string;
    baseQuantity: number;
    scalingFactor: number;
}

export interface DimensionTheme {
    name: string;
    visualStyle: string;
    musicMood: string;
    colorPalette: string[];
}

export interface Objective {
    id: string;
    description: string;
    objectiveType: 'score' | 'time' | 'collect' | 'defeat' | 'survive' | string;
    targetValue: number;
    isOptional: boolean;
}

export interface SessionResult {
    dimensionId: string;
    difficulty: number;
    playerLevel: number;
    score: number;
    durationSecs: number;
    completed: boolean;
}

// Re-export the 4 AI brains so consumers can pick whichever they need.
export { GameplayCombinerAI, CombinationSuggestion, GameplayStage } from './GameplayCombinerAI';
export { ContentGeneratorAI, ThemeContent, VisualStyle, MusicMood } from './ContentGeneratorAI';
export { SmartWorldAI, WorldEventDraft, WorldEventKind } from './SmartWorldAI';

export class DimensionGenerator {
    private rng: () => number;
    private adjectives: string[];
    private nouns: string[];
    private themes: string[];

    constructor(seed: number) {
        this.rng = this.createSeededRandom(seed);
        this.adjectives = [
            '混沌', '永恒', '幻影', '量子', '虚空', '烈焰', '冰霜',
            '雷霆', '暗影', '光辉', '深渊', '星辰', '时空', '命运',
        ];
        this.nouns = [
            '迷宫', '战场', '神殿', '深渊', '花园', '塔楼', '竞技场',
            '秘境', '次元', '试炼', '回廊', '领域', '裂隙', '梦境',
        ];
        this.themes = [
            '赛博朋克', '奇幻森林', '海底世界', '太空站', '古墓',
            '浮空城', '熔岩地带', '冰原', '沙漠绿洲', '暗黑地牢',
        ];
    }

    generate(config: GenerationConfig): DimensionBlueprint {
        const availableAtoms = this.getDefaultAtoms();
        const numAtoms = Math.floor(
            this.rng() * (config.maxAtoms - config.minAtoms + 1)
        ) + config.minAtoms;

        const selectedAtoms = this.selectRandomAtoms(availableAtoms, numAtoms);
        const difficulty = this.rng() * (config.difficultyRange[1] - config.difficultyRange[0]) + config.difficultyRange[0];
        const atomWeights = this.generateWeights(selectedAtoms);
        const rules = this.generateRules(selectedAtoms, difficulty);
        const rewards = this.generateRewards(selectedAtoms, difficulty, config.rewardMultiplier);
        const theme = this.generateTheme();
        const name = this.generateName();
        const objectives = this.generateObjectives(selectedAtoms, difficulty);

        return {
            id: `dim_${Date.now()}_${Math.random().toString(36).substr(2, 6)}`,
            name,
            description: `在${theme.name}中挑战${selectedAtoms.length}种玩法的组合`,
            atomIds: selectedAtoms,
            atomWeights,
            difficulty,
            rules,
            rewards,
            theme,
            timeLimitSecs: difficulty > 0.7 ? 180 : null,
            objectives,
        };
    }

    private getDefaultAtoms(): string[] {
        return [
            'match3',
            'tower_defense',
            'card',
            'turn_combat',
            'parkour',
            'puzzle',
            'shooting',
            'synthesis',
        ];
    }

    private selectRandomAtoms(available: string[], count: number): string[] {
        const shuffled = [...available].sort(() => this.rng() - 0.5);
        return shuffled.slice(0, count);
    }

    private generateWeights(atoms: string[]): Record<string, number> {
        const weights: Record<string, number> = {};
        const total = atoms.length;
        for (const atom of atoms) {
            weights[atom] = (0.5 + this.rng() * 0.5) / total;
        }
        return weights;
    }

    private generateRules(_atoms: string[], difficulty: number): GeneratedRule[] {
        const templates: Array<{ id: string; name: string; desc: string; type: GeneratedRule['ruleType'] }> = [
            { id: 'speed_boost', name: '加速', desc: '行动速度提升', type: 'modifier' },
            { id: 'double_score', name: '双倍得分', desc: '得分翻倍', type: 'modifier' },
            { id: 'resource_drain', name: '资源消耗', desc: '资源持续消耗', type: 'constraint' },
            { id: 'chain_bonus', name: '连锁奖励', desc: '连续操作获得额外奖励', type: 'trigger' },
            { id: 'time_pressure', name: '时间压力', desc: '倒计时加速', type: 'constraint' },
        ];

        const numRules = Math.min(
            Math.floor(1 + difficulty * 3),
            templates.length
        );

        const shuffled = templates.sort(() => this.rng() - 0.5);
        return shuffled.slice(0, numRules).map(t => ({
            ruleId: t.id,
            name: t.name,
            description: t.desc,
            ruleType: t.type,
            params: { intensity: difficulty },
        }));
    }

    private generateRewards(_atoms: string[], difficulty: number, multiplier: number): GeneratedReward[] {
        const baseGold = Math.floor(50 * difficulty * multiplier);
        const baseGem = Math.max(1, Math.floor(5 * difficulty * multiplier));

        const rewards: GeneratedReward[] = [
            { itemId: 'gold', baseQuantity: baseGold, scalingFactor: 1.0 + difficulty },
            { itemId: 'gem', baseQuantity: baseGem, scalingFactor: 0.5 + difficulty },
        ];

        if (difficulty > 0.6) {
            rewards.push({
                itemId: 'rare_chest',
                baseQuantity: 1,
                scalingFactor: difficulty,
            });
        }

        return rewards;
    }

    private generateTheme(): DimensionTheme {
        const adjIdx = Math.floor(this.rng() * this.adjectives.length);
        const themeIdx = Math.floor(this.rng() * this.themes.length);

        return {
            name: `${this.adjectives[adjIdx]}·${this.themes[themeIdx]}`,
            visualStyle: this.themes[themeIdx],
            musicMood: this.rng() > 0.5 ? 'epic' : 'mysterious',
            colorPalette: ['#FF6B6B', '#4ECDC4', '#45B7D1'],
        };
    }

    private generateName(): string {
        const adjIdx = Math.floor(this.rng() * this.adjectives.length);
        const nounIdx = Math.floor(this.rng() * this.nouns.length);
        return `${this.adjectives[adjIdx]}${this.nouns[nounIdx]}`;
    }

    private generateObjectives(atoms: string[], difficulty: number): Objective[] {
        const objectives: Objective[] = [
            {
                id: 'main_score',
                description: `达到${Math.floor(1000 * difficulty)}分`,
                objectiveType: 'score',
                targetValue: Math.floor(1000 * difficulty),
                isOptional: false,
            },
        ];

        if (atoms.length > 1) {
            objectives.push({
                id: 'combo_master',
                description: '完成3次组合连击',
                objectiveType: 'combo',
                targetValue: 3,
                isOptional: true,
            });
        }

        return objectives;
    }

    private createSeededRandom(seed: number): () => number {
        let s = seed;
        return () => {
            s = (s * 9301 + 49297) % 233280;
            return s / 233280;
        };
    }
}

export class BalanceTuner {
    private targetWinRate: number = 0.6;
    private history: SessionResult[] = [];

    recordResult(result: SessionResult): void {
        this.history.push(result);
    }

    suggestDifficulty(playerLevel: number): number {
        const recent = this.history
            .slice(-20)
            .filter(d => Math.abs(d.playerLevel - playerLevel) <= 2);

        if (recent.length === 0) {
            return 0.3 + playerLevel * 0.05;
        }

        const winRate = recent.filter(d => d.completed).length / recent.length;

        let adjustment = 0;
        if (winRate > this.targetWinRate + 0.1) {
            adjustment = 0.1;
        } else if (winRate < this.targetWinRate - 0.1) {
            adjustment = -0.1;
        }

        const base = 0.3 + playerLevel * 0.05;
        return Math.max(0.1, Math.min(1.0, base + adjustment));
    }
}

import { GameplayCombinerAI } from './GameplayCombinerAI';
import { ContentGeneratorAI } from './ContentGeneratorAI';
import { SmartWorldAI } from './SmartWorldAI';

export class AIEngine {
    /** A — 玩法组合 AI: pick the right combo of gameplay atoms for this player. */
    public gameplayAI: GameplayCombinerAI;
    /** B — 内容生成 AI: produce theme name, art prompt, BGM prompt, intro lore. */
    public contentAI: ContentGeneratorAI;
    /** C — 平衡 AI: tune difficulty from historical win/loss data. */
    public tuner: BalanceTuner;
    /** D — 智能 NPC / 世界 AI: roll transient world events and NPC dialogue. */
    public worldAI: SmartWorldAI;

    /** Internal generator of dimension blueprints (combines the 4 AIs' output). */
    private generator: DimensionGenerator;

    constructor(seed: number) {
        this.gameplayAI = new GameplayCombinerAI();
        this.contentAI = new ContentGeneratorAI(seed);
        this.tuner = new BalanceTuner();
        this.worldAI = new SmartWorldAI(seed);
        this.generator = new DimensionGenerator(seed);
    }

    generateDimension(config: GenerationConfig): DimensionBlueprint {
        // 1) Balance AI hands us a suggested difficulty band.
        const suggested = this.tuner.suggestDifficulty(config.playerLevel);
        const adjustedConfig: GenerationConfig = {
            ...config,
            difficultyRange: [
                Math.max(0.1, suggested - 0.1),
                Math.min(1.0, suggested + 0.1),
            ],
        };

        // 2) DimensionGenerator emits the structural blueprint.
        const blueprint = this.generator.generate(adjustedConfig);

        // 3) Content AI enriches the theme with lore / art / music prompts.
        const stage = this.gameplayAI.classifyStage(config.playerLevel);
        const theme = this.contentAI.generate(stage, blueprint.atomIds, blueprint.difficulty);
        blueprint.name = theme.themeName;
        blueprint.description = theme.introLore;
        blueprint.theme = {
            name: theme.themeName,
            visualStyle: theme.visualStyle,
            musicMood: theme.musicMood,
            colorPalette: theme.colorPalette,
        };

        return blueprint;
    }

    recordSession(result: SessionResult): void {
        this.tuner.recordResult(result);
    }

    suggestDifficulty(playerLevel: number): number {
        return this.tuner.suggestDifficulty(playerLevel);
    }
}
